packages/SwingSet/src/kernel/slogger.js

```diff
@@ -0,0 +1,111 @@
+/* global harden */
+import { assert } from '@agoric/assert';
+
+const IDLE = 'idle';
+const STARTUP = 'startup';
+const DELIVERY = 'delivery';
+
+export function makeDummySlogger(makeConsole) {
+  return harden({
+    addVat: () => 0,
+    vatConsole: () => makeConsole('disabled slogger'),
+    startup: () => () => 0, // returns nop finish() function
+    delivery: () => () => 0,
+    syscall: () => () => 0,
+  });
+}
+
+export function makeSlogger(writeObj) {
+  const write = writeObj ? e => writeObj(harden(e)) : () => 0;
+
+  const vatSlogs = new Map(); // vatID -> vatSlog
+
+  function makeVatSlog(vatID) {
+    let state = IDLE; // or STARTUP or DELIVERY
+    let crankNum;
+    let deliveryNum = 0;
+    let syscallNum;
+
+    function assertOldState(exp, msg) {
+      assert(state === exp, `vat ${vatID} in ${state}, not ${exp}: ${msg}`);
+    }
+
+    function vatConsole(origConsole) {
+      const vc = {};
+      for (const level of ['debug', 'log', 'info', 'warn', 'error']) {
+        vc[level] = (...args) => {
+          origConsole[level](...args);
+          const when = { state, crankNum, vatID, deliveryNum };
+          write({ type: 'console', ...when, level, args });
+        };
+      }
+      return harden(vc);
+    }
+
+    function startup() {
+      // provide a context for console calls during startup
+      assertOldState(IDLE, 'did startup get called twice?');
+      state = STARTUP;
+      function finish() {
+        assertOldState(STARTUP, 'startup-finish called twice?');
+        state = IDLE;
+      }
+      return harden(finish);
+    }
+
+    // kd: kernelDelivery, vd: vatDelivery
+    function delivery(newCrankNum, kd, vd) {
```

Comment:
I don't understand logging both `kd` and `vd`.  Since one is a translation of the other, I would think that the only reason to log them both is if you are debugging the translation mechanism itself, which is unusual now and will be almost unheard of later.  Doing it universally and perpetually seems like a lot of overhead to carry, especially in production.

Reply:
My goal is to give the rendering tool (which has access to the slog data, but not the HostDB, which might be far away or long-gone by now, e.g. a dynamic vat might have been deleted) enough data to correlate cross-vat messages, as well as correlate deliveries/syscalls with log entries created by e.g. liveslots. So I need at least the `slots` in both kernel form and vat form.

It'd be sufficient (more work here, less data stored in the slog file) to store just the KernelDelivery record, but also parse the record to find all the slots (message target, message result, message argument slots, promise resolution target, promise resolution presence, promise resolution/rejection data slots), translate them (again) (and without accidentally allocating anything) (and without re-retiring anything), and record a table mapping all those slots to their vat-side equivalents.

I figured it was easier and less dangerous (in the face of retiring promises) to just record both the pre- and post- translation records.